File: tools/difftest_A.py
# capture function-level differential test pairs for part A's pure functions
import sys, os, json, zlib, struct
from emu_trace import TEmu, BASE32
from unicorn.x86_const import *
OUT = 'fxA_tests'; os.makedirs(OUT, exist_ok=True)
WANT = {0x1545e: 4, 0x15508: 4, 0x1570b: 3, 0x1574e: 2, 0x15660: 2, 0x155b3: 1, 0x15646: 1, 0x156ff: 2, 0xe1cc: 1, 0xe27a: 3}
hits = {k: 0 for k in WANT}
e = TEmu(cmdline="", instr_per_sec=30_000_000)
pending = {}
def mk(off):
    def cb(em):
        if hits[off] >= WANT[off]: return
        n = hits[off]; hits[off] += 1
        regs = em.regs(); ret = em.ret_addr()
        before = em.snapshot()
        tag = "%x_%d" % (off, n)
        open("%s/%s_before.bin.z" % (OUT, tag), 'wb').write(zlib.compress(before, 1))
        def after(em2):
            r2 = em2.regs()
            open("%s/%s_after.bin.z" % (OUT, tag), 'wb').write(zlib.compress(em2.snapshot(), 1))
            json.dump(dict(fn=off, regs_in=regs, regs_out=r2, ret=ret), open("%s/%s.json" % (OUT, tag), 'w'))
            print("captured", tag, flush=True)
        em.bp_once(ret, after)
    return cb
for off in WANT: e.bp(off, mk(off))
e.run(max_seconds=8.0, until=lambda em: all(hits[k] >= WANT[k] for k in WANT))
print(hits)

File: tools/difftest_A.js
// node difftest_A.js [DIR]  — run captured (before, regs) -> (after) pairs through the JS port of part A's functions
'use strict';
const fs = require('fs'), zlib = require('zlib'), path = require('path');
require('./hplus_core.js'); require('./hplus_main.js'); require('./hplus_fxA.js');
const HP = globalThis.HP;
const DIR = process.argv[2] || 'fxA_tests';
const image = HP.loadImageNode('image32.bin');
const CALL = {
  0x1545e: r => HP.fn_1545e(r.edx, r.ebp, r.esi, r.edi),
  0x15508: r => HP.fn_15508(r.edx, r.ebp, r.edi),
  0x1570b: r => HP.fn_1570b(r.esi, r.ebp),
  0x1574e: r => HP.fn_1574e(r.ebp, r.edi),
  0x15660: r => ({ edi: HP.fn_15660(r.eax, r.ebx, r.edi) }),
  0x155b3: r => ({ edi: HP.fn_155b3(r.edi) }),
  0x15646: r => ({ edi: HP.fn_15646(r.edi) }),
  0x156ff: r => ({ edi: HP.fn_156ff(r.esi, r.edi) }),
  0xe1cc: r => HP.fn_e1cc(),
  0xe27a: r => HP.fn_e27a(r.ebx, r.ecx, r.edx),
};
let pass = 0, fail = 0;
for (const f of fs.readdirSync(DIR).filter(x => x.endsWith('.json')).sort()) {
  const t = JSON.parse(fs.readFileSync(path.join(DIR, f)));
  const before = zlib.inflateSync(fs.readFileSync(path.join(DIR, f.replace('.json', '_before.bin.z'))));
  const after = zlib.inflateSync(fs.readFileSync(path.join(DIR, f.replace('.json', '_after.bin.z'))));
  HP.init(image); HP.M.set(before, 0);
  const ret = CALL[t.fn](t.regs_in) || {};
  const M = HP.M; const n = Math.min(after.length, M.length);
  const volatile = (a) => (a >= 0x32800 && a < 0x32c20) || (a >= 0x2ee50 && a < 0x2ee64) || (a >= 0x2ee3d && a < 0x2ee4d) || (a >= 0x3f090 && a < 0x5f090) || (a >= 0x1283b0 && a < 0x1336f0) || (a >= 0x1336f0 - 0x3c00 && a < 0x1336f0);
  let ranges = [], start = -1, count = 0;
  for (let i = 0; i < n; i++) {
    const d = M[i] !== after[i] && !volatile(i);
    if (d && start < 0) start = i;
    if (!d && start >= 0) { ranges.push([start, i]); start = -1; }
    if (d) count++;
  }
  if (start >= 0) ranges.push([start, n]);
  let regok = true;
  for (const k of Object.keys(ret)) if ((ret[k] >>> 0) !== (t.regs_out[k] >>> 0)) regok = false;
  const ok = count === 0 && regok;
  if (ok) pass++; else fail++;
  console.log(`${f}: ${ok ? 'OK' : 'FAIL'} (${count} bytes differ${ranges.length ? ': ' + ranges.slice(0, 8).map(r => r[0].toString(16) + '-' + r[1].toString(16)).join(' ') : ''}${regok ? '' : '; return regs differ ' + JSON.stringify(ret) + ' vs ' + JSON.stringify(t.regs_out)})`);
}
console.log(`pass ${pass} fail ${fail}`);
